### PR TITLE
update min Java version, Maven dependencies and plugins

### PR DESCRIPTION
**Pull request by don-vip**

### CODE REVIEW

**Comment by don-vip** [comment]
Travis build if failing for OpenJDK7 only. Do you want to keep it, or could we drop it and keep only support of OpenJDK8+?

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java [hunk 1]
```diff
@@ -383,6 +384,6 @@ public void onLoad(Run<?, ?> run) {
 
 	@Override
 	public Collection<? extends Action> getProjectActions() {
-		return jacocoProjectActions;
+		return jacocoProjectAction != null ? Arrays.asList(jacocoProjectAction) : Collections.emptyList();
```
Comment (centic9) [inline src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: I would use Collections.singletonList() here instead of Arrays.asList()

File: src/main/java/hudson/plugins/jacoco/ExecutionFileLoader.java [hunk 2]
```diff
@@ -33,10 +34,10 @@
 		private String[] includes;
 		private String[] excludes;
 		
-		private ExecutionDataStore executionDataStore;
-		private SessionInfoStore sessionInfoStore;
+		private transient ExecutionDataStore executionDataStore;
```
Comment (centic9) [inline src/main/java/hudson/plugins/jacoco/ExecutionFileLoader.java]: Can you add a comment why "transient" is necessary/useful here?

File: pom.xml [hunk 3]
```diff
@@ -97,7 +97,10 @@ THE SOFTWARE.
       <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
       <project.build.outputEncoding>UTF-8</project.build.outputEncoding>
       <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+      <ant.version>1.9.2</ant.version>
```
Comment (centic9) [inline pom.xml]: Why do we need a dependency on Ant now?

File: pom.xml [hunk 4]
```diff
@@ -128,6 +131,23 @@ THE SOFTWARE.
...
+	  </dependency>
+      <dependency>
+	    <groupId>commons-io</groupId>
+	    <artifactId>commons-io</artifactId>
+	    <version>2.5</version>
+	    <scope>runtime</scope>
+	  </dependency>
+      <dependency>
+	    <groupId>xerces</groupId>
+	    <artifactId>xercesImpl</artifactId>
```
Comment (centic9) [inline pom.xml]: Xerces dependency should not be needed since Java 7, why is it necessary here?

File: pom.xml [hunk 5]
```diff
@@ -128,6 +131,23 @@ THE SOFTWARE.
           </exclusion>
         </exclusions> 
       </dependency>
+      <dependency>
+	    <groupId>org.apache.ant</groupId>
+	    <artifactId>ant</artifactId>
+	    <version>${ant.version}</version>
+	  </dependency>
+      <dependency>
+	    <groupId>commons-io</groupId>
```
Comment (centic9) [inline pom.xml]: I don't see any usages of commons-io in the changed code, why is the dependency necessary now?

File: pom.xml [hunk 6]
```diff
@@ -195,30 +215,76 @@ THE SOFTWARE.
             <version>3.5</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.apache.maven.reporting</groupId>
```
Comment (centic9) [inline pom.xml]: Why do we need all these Maven dependencies now? Where do we actually use functionality from them?

File: src/main/java/hudson/plugins/jacoco/portlet/JacocoLoadData.java [hunk 7]
```diff
@@ -137,7 +131,7 @@ private static void summarize(Map<String, JacocoCoverageResultSummary> summaries
 
     JacocoCoverageResultSummary jacocoCoverageResult = getResult(run);
 	
-	String date = DATE_FORMAT.format(runDate.getTime());
+	String date = new SimpleDateFormat("yyyy-MM-dd").format(runDate.getTime());
```
Comment (centic9) [inline src/main/java/hudson/plugins/jacoco/portlet/JacocoLoadData.java]: Unfortunately you are mixing many different changes here, not only Java 7->8, these seem to be intended to improve concurrent behavior somehow. This makes review a bit harder, would in general be better to send separate PRs for these unrelated changes.

File: pom.xml (hunk 6 above)
Comment (don-vip) [reply pom.xml]: Because of the switch to org.jenkins-ci.plugins 3.0, which defines a very strict maven enforcer configuration. If not met, the build fails. The enforcer rule causing all of this is "[Require Upper Bound Dependencies](https://maven.apache.org/enforcer/enforcer-rules/requireUpperBoundDeps.html)".

File: src/main/java/hudson/plugins/jacoco/ExecutionFileLoader.java (hunk 2 above)
Comment (don-vip) [reply src/main/java/hudson/plugins/jacoco/ExecutionFileLoader.java]: Sure. It's because of the switch to org.jenkins-ci.plugins 3.0, which defines a FindBugs analysis.

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 1 above)
Comment (don-vip) [reply src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: ok

File: src/main/java/hudson/plugins/jacoco/portlet/JacocoLoadData.java (hunk 7 above)
Comment (don-vip) [reply src/main/java/hudson/plugins/jacoco/portlet/JacocoLoadData.java]: This is not unrelated. Yet again caused by the switch to org.jenkins-ci.plugins 3.0 and the underlying FindBugs analysis. If not fixed, the build fails.

**Comment by don-vip** [comment]
This new enforcer rule is annoying. For example, I am trying to update to parent pom 3.2, and this is what I get:
```
[WARNING] Rule 6: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for org.ow2.asm:asm:6.0 paths to dependency are:
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.ow2.asm:asm:6.0
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.ow2.asm:asm-commons:6.0
    +-org.ow2.asm:asm-tree:6.0
      +-org.ow2.asm:asm:6.0
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.apache.maven.plugins:maven-plugin-plugin:3.5
    +-org.apache.maven.plugin-tools:maven-plugin-tools-generators:3.5
      +-org.ow2.asm:asm:6.0_ALPHA
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.apache.maven.plugins:maven-plugin-plugin:3.5
    +-org.apache.maven.plugin-tools:maven-plugin-tools-annotations:3.5
      +-org.ow2.asm:asm:6.0_ALPHA
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.jenkins-ci.main:jenkins-core:2.54
    +-com.github.jnr:jnr-posix:3.0.1
      +-com.github.jnr:jnr-ffi:1.0.7
        +-org.ow2.asm:asm:4.0
,
Require upper bound dependencies error for org.ow2.asm:asm-commons:6.0 paths to dependency are:
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.ow2.asm:asm-commons:6.0
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.apache.maven.plugins:maven-plugin-plugin:3.5
    +-org.apache.maven.plugin-tools:maven-plugin-tools-generators:3.5
      +-org.ow2.asm:asm-commons:6.0_ALPHA
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.jenkins-ci.main:jenkins-core:2.54
    +-com.github.jnr:jnr-posix:3.0.1
      +-com.github.jnr:jnr-ffi:1.0.7
        +-org.ow2.asm:asm-commons:4.0
]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 11.768 s
[INFO] Finished at: 2017-12-05T23:04:09+01:00
[INFO] Final Memory: 39M/130M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (display-info) on project jacoco: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
```

**Comment by kscaldef** [comment]
What are the prospects for moving this along?  I was making a pass at updating to the newly released JaCoCo 0.8 version it appears that requiring JDK 1.7 compatibility is holding it back

**Comment by don-vip** [comment]
I'm still waiting for https://github.com/apache/maven-plugin-tools/pull/9 to be merged, I didn't think it would take this long.

File: pom.xml (hunk 3 above)
Comment (don-vip) [reply pom.xml]: It was to fix this error, but it's probably not the best way:

```
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (display-info) @ jacoco ---
[INFO] Ignoring requireUpperBoundDeps in org.kohsuke:access-modifier-annotation
[WARNING] Rule 6: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for org.apache.ant:ant:1.8.4 paths to dependency are:
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.jenkins-ci.main:jenkins-core:2.54
    +-org.apache.ant:ant:1.8.4
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.jenkins-ci.plugins:dashboard-view:2.9.11
    +-org.jenkins-ci.main:maven-plugin:2.13
      +-org.jenkins-ci.main.maven:maven-agent:1.7
        +-org.apache.ant:ant:1.8.4
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.jenkins-ci.plugins:dashboard-view:2.9.11
    +-org.jenkins-ci.main:maven-plugin:2.13
      +-org.jenkins-ci.lib:lib-jenkins-maven-embedder:3.11
        +-org.apache.ant:ant:1.9.2
```

File: pom.xml (hunk 5 above)
Comment (don-vip) [reply pom.xml]: Same reason than ant, but not needed anymore after a new update

File: pom.xml (hunk 4 above)
Comment (don-vip) [reply pom.xml]: To fix this error but I should find another way:

```
[INFO] --- maven-enforcer-plugin:3.0.0-M1:enforce (display-info) @ jacoco ---
[INFO] Ignoring requireUpperBoundDeps in org.kohsuke:access-modifier-annotation
[WARNING] Rule 6: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
Failed while enforcing RequireUpperBoundDeps. The error(s) are [
Require upper bound dependencies error for xerces:xercesImpl:2.9.1 paths to dependency are:
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.apache.maven.reporting:maven-reporting-impl:2.3
    +-org.apache.maven.doxia:doxia-core:1.2
      +-xerces:xercesImpl:2.9.1
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.jenkins-ci.main:jenkins-test-harness:2.32
    +-org.jenkins-ci.main:jenkins-test-harness-htmlunit:2.18-1
      +-xerces:xercesImpl:2.11.0
and
+-org.jenkins-ci.plugins:jacoco:2.3-SNAPSHOT
  +-org.jenkins-ci.main:jenkins-test-harness:2.32
    +-org.jenkins-ci.main:jenkins-test-harness-htmlunit:2.18-1
      +-net.sourceforge.nekohtml:nekohtml:1.9.22
        +-xerces:xercesImpl:2.11.0
```

File: src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java (hunk 1 above)
Comment (don-vip) [reply src/main/java/hudson/plugins/jacoco/JacocoBuildAction.java]: done

**Comment by don-vip** [comment]
OK it appears the other PR in maven-plugin-tools is not needed anymore, no need to wait for it.
@centic9 : can you please check if everything's OK now?

**Comment by centic9** [comment]
Looks good now, I expect we will release a major version due to the changes in dependencies.

**Comment by kscaldef** [comment]
@centic9 any idea when that release will happen?

**Comment by centic9** [comment]
I tried to release, but as usual the release-steps fail at some point with some permission denied. It seems I am not able to get this solved in a way that works longer than for one release, grrr.... anybody else who knows what to do?

 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:2.8.2:deploy (default-deploy) on project jacoco: Failed to deploy artifacts: Could not transfer artifact org.jenkins-ci.plugins:jacoco:hpi:3.0 from/to maven.jenkins-ci.org (https://repo.jenkins-ci.org/releases): Access denied to: https://repo.jenkins-ci.org/releases/org/jenkins-ci/plugins/jacoco/3.0/jacoco-3.0.hpi, ReasonPhrase: . -> [Help 1]

I re-created the settings.xml and checked other things, but to no avail ...
